src/stream/src/executor/over_window/over_partition.rs

```diff
@@ -0,0 +1,1256 @@
+        Vec<(
+            &'cache CacheKey,
+            &'cache CacheKey,
+            &'cache CacheKey,
+            &'cache CacheKey,
+        )>,
+    )>
+    where
+        's: 'cache,
+    {
+        let delta_first = delta.first_key_value().unwrap().0.as_normal_expect();
+        let delta_last = delta.last_key_value().unwrap().0.as_normal_expect();
+
+        if self.has_unbounded_frame {
+            // for unbounded frame, we finally need all entries of the partition in the cache, so
+            // for simplicity we just load all entries at once
+            self.extend_cache_to_boundary(table).await?;
+        } else {
+            // ensure the cache covers all delta (if possible)
+            self.extend_cache_by_range(table, delta_first..=delta_last)
+                .await?;
+        }
+
+        loop {
+            let (left_reached_sentinel, right_reached_sentinel) = {
+                // SAFETY: Here we shortly borrow the range cache and turn the reference into a
+                // `'cache` one to bypass the borrow checker. This is safe because we only return
+                // the reference once we don't need to do any further mutation.
+                let cache_inner = unsafe { &*(self.range_cache.inner() as *const _) };
+                let ranges =
+                    self::find_affected_ranges(self.calls, DeltaBTreeMap::new(cache_inner, delta));
+
+                if ranges.is_empty() {
+                    // no ranges affected, we're done
+                    return Ok((DeltaBTreeMap::new(cache_inner, delta), ranges));
+                }
+
+                let left_reached_sentinel = ranges.first().unwrap().0.is_sentinel();
+                let right_reached_sentinel = ranges.last().unwrap().3.is_sentinel();
+
+                if !left_reached_sentinel && !right_reached_sentinel {
+                    // all affected ranges are already cached, we're done
+                    return Ok((DeltaBTreeMap::new(cache_inner, delta), ranges));
+                }
+
+                (left_reached_sentinel, right_reached_sentinel)
+            };
+
+            if left_reached_sentinel {
+                // TODO(rc): should count cache miss for this, and also the below
+                tracing::debug!(partition=?self.this_partition_key, "partition cache left extension triggered");
+                let left_most = self.cache_real_first_key().unwrap_or(delta_first).clone();
+                self.extend_cache_leftward_by_n(table, &left_most).await?;
+            }
+            if right_reached_sentinel {
+                tracing::debug!(partition=?self.this_partition_key, "partition cache right extension triggered");
+                let right_most = self.cache_real_last_key().unwrap_or(delta_last).clone();
+                self.extend_cache_rightward_by_n(table, &right_most).await?;
+            }
+            tracing::debug!(partition=?self.this_partition_key, "partition cache extended");
+        }
```

Comment:
Terminateability: `extend_cache_leftward_by_n` and `extend_cache_rightward_by_n` keep pushing the cache to the boundary of current partition. In these two methods, when any side of boundary is reached, the sentinel key will be removed, so finally `self::find_affected_ranges` will return ranges without any sentinels.